File: services/kotlin/src/main/kotlin/com/corestack/repository/ProjectRepo.kt
package com.corestack.repository

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.jdbi.v3.core.Jdbi
import java.time.Instant
import java.util.UUID

data class Project(
    val id: UUID,
    val workspaceId: UUID,
    val name: String,
    val description: String,
    val status: String,
    val createdAt: Instant,
    val updatedAt: Instant,
    val deletedAt: Instant? = null
)

data class CreateProjectParams(val workspaceId: UUID, val name: String, val description: String)
data class UpdateProjectParams(val id: UUID, val name: String, val description: String, val status: String)
data class ListProjectsParams(val workspaceId: UUID, val pageSize: Int = 0, val pageToken: String = "")
data class ProjectList(val projects: List<Project>, val nextPageToken: String, val totalCount: Int)

class ProjectRepo(private val jdbi: Jdbi) {

    suspend fun create(params: CreateProjectParams): Project = withContext(Dispatchers.IO) {
        jdbi.withHandle<Project, Exception> { handle ->
            handle.createQuery(
                """INSERT INTO projects (id, workspace_id, name, description, status, created_at, updated_at)
                   VALUES (gen_random_uuid(), :workspaceId, :name, :description, 'active', NOW(), NOW())
                   RETURNING id, workspace_id, name, description, status, created_at, updated_at, deleted_at"""
            )
                .bind("workspaceId", params.workspaceId)
                .bind("name", params.name)
                .bind("description", params.description)
                .map { rs, _ ->
                    Project(
                        id = rs.getObject("id", UUID::class.java),
                        workspaceId = rs.getObject("workspace_id", UUID::class.java),
                        name = rs.getString("name"),
                        description = rs.getString("description"),
                        status = rs.getString("status"),
                        createdAt = rs.getTimestamp("created_at").toInstant(),
                        updatedAt = rs.getTimestamp("updated_at").toInstant(),
                        deletedAt = rs.getTimestamp("deleted_at")?.toInstant()
                    )
                }
                .first()
        }
    }

    suspend fun getById(id: UUID): Project = withContext(Dispatchers.IO) {
        jdbi.withHandle<Project?, Exception> { handle ->
            handle.createQuery(
                """SELECT id, workspace_id, name, description, status, created_at, updated_at, deleted_at
                   FROM projects WHERE id = :id AND deleted_at IS NULL"""
            )
                .bind("id", id)
                .map { rs, _ ->
                    Project(
                        id = rs.getObject("id", UUID::class.java),
                        workspaceId = rs.getObject("workspace_id", UUID::class.java),
                        name = rs.getString("name"),
                        description = rs.getString("description"),
                        status = rs.getString("status"),
                        createdAt = rs.getTimestamp("created_at").toInstant(),
                        updatedAt = rs.getTimestamp("updated_at").toInstant(),
                        deletedAt = rs.getTimestamp("deleted_at")?.toInstant()
                    )
                }
                .firstOrNull()
        } ?: throw NotFoundError("project not found")
    }

    suspend fun list(params: ListProjectsParams): ProjectList = withContext(Dispatchers.IO) {
        val pageSize = if (params.pageSize in 1..100) params.pageSize else 20

        jdbi.withHandle<ProjectList, Exception> { handle ->
            val totalCount = handle.createQuery(
                "SELECT COUNT(*)::int FROM projects WHERE workspace_id = :workspaceId AND deleted_at IS NULL"
            )
                .bind("workspaceId", params.workspaceId)
                .mapTo(Int::class.java)
                .first()

            val query = if (params.pageToken.isNotEmpty()) {
                val cursorId = UUID.fromString(params.pageToken)
                handle.createQuery(
                    """SELECT id, workspace_id, name, description, status, created_at, updated_at, deleted_at
                       FROM projects WHERE workspace_id = :workspaceId AND deleted_at IS NULL AND id < :cursorId
                       ORDER BY created_at DESC, id DESC LIMIT :limit"""
                )
                    .bind("workspaceId", params.workspaceId)
                    .bind("cursorId", cursorId)
                    .bind("limit", pageSize + 1)
            } else {
                handle.createQuery(
                    """SELECT id, workspace_id, name, description, status, created_at, updated_at, deleted_at
                       FROM projects WHERE workspace_id = :workspaceId AND deleted_at IS NULL
                       ORDER BY created_at DESC, id DESC LIMIT :limit"""
                )
                    .bind("workspaceId", params.workspaceId)
                    .bind("limit", pageSize + 1)
            }

            val projects = query.map { rs, _ ->
                Project(
                    id = rs.getObject("id", UUID::class.java),
                    workspaceId = rs.getObject("workspace_id", UUID::class.java),
                    name = rs.getString("name"),
                    description = rs.getString("description"),
                    status = rs.getString("status"),
                    createdAt = rs.getTimestamp("created_at").toInstant(),
                    updatedAt = rs.getTimestamp("updated_at").toInstant(),
                    deletedAt = rs.getTimestamp("deleted_at")?.toInstant()
                )
            }.toMutableList()

            val nextPageToken = if (projects.size > pageSize) {
                val token = projects[pageSize].id.toString()
                projects.removeAt(pageSize)
                token
            } else ""

            ProjectList(projects = projects, nextPageToken = nextPageToken, totalCount = totalCount)
        }
    }

    suspend fun update(params: UpdateProjectParams): Project = withContext(Dispatchers.IO) {
        jdbi.withHandle<Project?, Exception> { handle ->
            handle.createQuery(
                """UPDATE projects SET name = :name, description = :description, status = :status, updated_at = NOW()
                   WHERE id = :id AND deleted_at IS NULL
                   RETURNING id, workspace_id, name, description, status, created_at, updated_at, deleted_at"""
            )
                .bind("name", params.name)
                .bind("description", params.description)
                .bind("status", params.status)
                .bind("id", params.id)
                .map { rs, _ ->
                    Project(
                        id = rs.getObject("id", UUID::class.java),
                        workspaceId = rs.getObject("workspace_id", UUID::class.java),
                        name = rs.getString("name"),
                        description = rs.getString("description"),
                        status = rs.getString("status"),
                        createdAt = rs.getTimestamp("created_at").toInstant(),
                        updatedAt = rs.getTimestamp("updated_at").toInstant(),
                        deletedAt = rs.getTimestamp("deleted_at")?.toInstant()
                    )
                }
                .firstOrNull()
        } ?: throw NotFoundError("project not found")
    }

    suspend fun delete(id: UUID): Unit = withContext(Dispatchers.IO) {
        val rowsAffected = jdbi.withHandle<Int, Exception> { handle ->
            handle.createUpdate(
                """UPDATE projects SET deleted_at = NOW(), updated_at = NOW()
                   WHERE id = :id AND deleted_at IS NULL"""
            )
                .bind("id", id)
                .execute()
        }
        if (rowsAffected == 0) throw NotFoundError("project not found")
    }
}

File: services/typescript/src/repository/project-repo.ts
import type postgres from "postgres";
import { NotFoundError } from "./errors.js";

export interface Project {
  id: string;
  workspaceId: string;
  name: string;
  description: string;
  status: string;
  createdAt: Date;
  updatedAt: Date;
  deletedAt: Date | null;
}

export interface CreateProjectParams {
  workspaceId: string;
  name: string;
  description: string;
}

export interface UpdateProjectParams {
  id: string;
  name: string;
  description: string;
  status: string;
}

export interface ListProjectsParams {
  workspaceId: string;
  pageSize: number;
  pageToken: string;
}

export interface ProjectList {
  projects: Project[];
  nextPageToken: string;
  totalCount: number;
}

export class ProjectRepo {
  constructor(private sql: postgres.Sql) {}

  async create(params: CreateProjectParams): Promise<Project> {
    const [row] = await this.sql<Project[]>`
      INSERT INTO projects (id, workspace_id, name, description, status, created_at, updated_at)
      VALUES (gen_random_uuid(), ${params.workspaceId}, ${params.name}, ${params.description}, 'active', NOW(), NOW())
      RETURNING id, workspace_id AS "workspaceId", name, description, status,
                created_at AS "createdAt", updated_at AS "updatedAt", deleted_at AS "deletedAt"
    `;
    return row;
  }

  async getById(id: string): Promise<Project> {
    const [row] = await this.sql<Project[]>`
      SELECT id, workspace_id AS "workspaceId", name, description, status,
             created_at AS "createdAt", updated_at AS "updatedAt", deleted_at AS "deletedAt"
      FROM projects WHERE id = ${id} AND deleted_at IS NULL
    `;
    if (!row) {
      throw new NotFoundError();
    }
    return row;
  }

  async list(params: ListProjectsParams): Promise<ProjectList> {
    let pageSize = params.pageSize;
    if (pageSize <= 0 || pageSize > 100) {
      pageSize = 20;
    }

    const [{ count }] = await this.sql<[{ count: number }]>`
      SELECT COUNT(*)::int AS count FROM projects
      WHERE workspace_id = ${params.workspaceId} AND deleted_at IS NULL
    `;

    let projects: Project[];
    if (params.pageToken) {
      projects = await this.sql<Project[]>`
        SELECT id, workspace_id AS "workspaceId", name, description, status,
               created_at AS "createdAt", updated_at AS "updatedAt", deleted_at AS "deletedAt"
        FROM projects WHERE workspace_id = ${params.workspaceId} AND deleted_at IS NULL AND id < ${params.pageToken}
        ORDER BY created_at DESC, id DESC LIMIT ${pageSize + 1}
      `;
    } else {
      projects = await this.sql<Project[]>`
        SELECT id, workspace_id AS "workspaceId", name, description, status,
               created_at AS "createdAt", updated_at AS "updatedAt", deleted_at AS "deletedAt"
        FROM projects WHERE workspace_id = ${params.workspaceId} AND deleted_at IS NULL
        ORDER BY created_at DESC, id DESC LIMIT ${pageSize + 1}
      `;
    }

    let nextPageToken = "";
    if (projects.length > pageSize) {
      nextPageToken = projects[pageSize].id;
      projects = projects.slice(0, pageSize);
    }

    return { projects, nextPageToken, totalCount: count };
  }

  async update(params: UpdateProjectParams): Promise<Project> {
    const [row] = await this.sql<Project[]>`
      UPDATE projects SET name = ${params.name}, description = ${params.description},
             status = ${params.status}, updated_at = NOW()
      WHERE id = ${params.id} AND deleted_at IS NULL
      RETURNING id, workspace_id AS "workspaceId", name, description, status,
                created_at AS "createdAt", updated_at AS "updatedAt", deleted_at AS "deletedAt"
    `;
    if (!row) {
      throw new NotFoundError();
    }
    return row;
  }

  async delete(id: string): Promise<void> {
    const result = await this.sql`
      UPDATE projects SET deleted_at = NOW(), updated_at = NOW()
      WHERE id = ${id} AND deleted_at IS NULL
    `;
    if (result.count === 0) {
      throw new NotFoundError();
    }
  }
}
